### PR TITLE
Use better buffer smart-ptrs

### PR DESCRIPTION
**Pull request by Desour**
This is a maintenance/refactor PR.

We have smart-ptrs for buffers, i.e. SharedBuffer<T>. These are mainly used for network stuff.

Improvements:
* Less unnecessary copying and value-initialization (aka 0-initialization).
* Types bigger than 1 byte work. (There were `memcpy`s that didn't take the type's size into account.)
* Values are copied using copy constructor (as opposed to forced memcpy).
* Copies are explicit: `some_fun(buffer.copy())` instead of `some_fun(buffer)`
* There's a `View<T>` class, and `ConstView<T> = View<const T>`, for borrowing buffer data.
* `Buffer<T>` is replaced with `UniqueBuffer<T>`, which is a more descriptive and grep-able name. It also does things different than `Buffer<T>`, it's essentially a `unique_ptr<T[]>` with a size.
* There's also a `make_unique_for_override` function (like the C++20 thing).
* `ConstSharedPtr<T>` is removed. It was just essentially a `shared_ptr<const T>` and not worth keeping.
* There are unittests.

## To do

This PR is a Ready for Review.

## How to test

* Compile.
* Run the unittests: `$ minetest_dev --run-unittests`
* Look at the code.
* Look at util/pointer.h first.

### CODE REVIEW

**Comment by Desour** [comment]
Added unittests and ConstView<>s.

In case you're getting uaf warnings from gcc12 for SharedBuffer, I'm pretty sure that's false positives. A small web-search also reveals that there are also other people experiencing uaf warning false positives.

**Comment by numberZero** [comment]
Did you consider using e.g. [span-lite](https://github.com/martinmoene/span-lite) instead of writing own `View` classes?

**Comment by Desour** [comment]
> Did you consider using e.g. [span-lite](https://github.com/martinmoene/span-lite) instead of writing own `View` classes?

I didn't. After all, the View class is very simple.
Also, doing it like this worked out quite nicely, i.e. I like how `View::copy` kind of closes the circle:
unique_buffer
(arrows are conversions; transitive edges left out; owning buffer<const>s also left out; black = costless; blue = allocates refcounter; yellow = move or copy; red = copy; black and blue are implicit)
Of course something similar would also work with `span`. But if we still want explicit `copy()` calls for copying, we'd have a different API if we copy from a span, i.e. `unique_buf.copy()` and `shared_buf.copy()` (and `some_view.copy()`) vs. `copy_to_unique_buf(some_span)`.

Anyway, I'm in general not against using such implementations of future C++ std utilities. But if we add one, we should add multiple at once, i.e. also [string-view-lite](https://github.com/martinmoene/string-view-lite) and similar.
(These particular libs are boost licensed, so I wonder whether by including them one already commits [the sin of using boost](https://dev.minetest.net/Code_style_guidelines#Use_STL.2C_avoid_Irrlicht_containers.2C_and_no.2C_Boost_will_not_even_be_considered.2C_so_forget_it).)

File: src/util/pointer.h [hunk 1]
```diff
@@ -22,126 +22,204 @@ with this program; if not, write to the Free Software Foundation, Inc.,
...
 
-private:
-	std::shared_ptr<T> ptr;
-};
+// Use this for default-initialization (aka don't fill with zeros)
+template <typename T, std::enable_if_t<!std::is_array<T>::value, bool> = true>
+std::unique_ptr<T> make_unique_for_overwrite()
+{
+	return std::unique_ptr<T>(new T);
+}
```
Comment (SmallJoker) [inline src/util/pointer.h:51]: What's the benefit of a separate function rather than using it as a constructor for UniqueBuffer?

**Comment by SmallJoker** [comment]
What is the advantage of `View` compared to `SharedBuffer<T>` references? It looks a bit overengineered for the functionalities that we need.

File: src/util/pointer.h (hunk 1 above)
Comment (Desour) [reply src/util/pointer.h:51]: This is just `std::make_unique_for_overwrite` from C++17.

I guess you wanted to comment on `make_buffer` or `make_buffer_for_overwrite`.
One benefit is the similarity to the std lib (=> more uniformity). Also, you can decide to default- or value-init.

**Comment by Desour** [comment]
> What is the advantage of `View` compared to `SharedBuffer<T>` references or `using View<T> = std::shared_ptr<T[]>;` ? It looks a bit overengineered for the functionalities that we need.

`View` can also refer to sub-spans of buffers. And it is not owning (as opposed to `std::shared_ptr<T[]>` (that one would also not make its size available)).

(In case you wonder why `SharedBuffer<T>` doesn't use a `std::shared_ptr<T[]>`: You can't convert a `shared_ptr` to a `unique_ptr` without copying, which is something this PR uses.)

Anyway, you can consider this PR low priority.

**Pull request closed without merging**